frame/support/procedural/src/benchmark.rs

```diff
@@ -0,0 +1,860 @@
+			setup_stmts: Vec::from(&item_fn.block.stmts[0..i]),
+			call_def,
+			verify_stmts: Vec::from(&item_fn.block.stmts[(i + 1)..item_fn.block.stmts.len()]),
+			extra,
+			skip_meta,
+		})
+	}
+}
+
+/// Parses and expands a `#[benchmarks]` or `#[instance_benchmarks]` invocation
+pub fn benchmarks(
+	attrs: TokenStream,
+	tokens: TokenStream,
+	instance: bool,
+) -> syn::Result<TokenStream> {
+	// gather module info
+	let module: ItemMod = syn::parse(tokens)?;
+	let mod_span = module.span();
+	let where_clause = match syn::parse::<Nothing>(attrs.clone()) {
+		Ok(_) => quote!(),
+		Err(_) => syn::parse::<WhereClause>(attrs)?.predicates.to_token_stream(),
+	};
+	let mod_vis = module.vis;
+	let mod_name = module.ident;
+
+	// consume #[benchmarks] attribute by exclusing it from mod_attrs
+	let mod_attrs: Vec<&Attribute> = module
+		.attrs
+		.iter()
+		.filter(|attr| syn::parse2::<keywords::benchmarks>(attr.to_token_stream()).is_err())
+		.collect();
+
+	let mut benchmark_names: Vec<Ident> = Vec::new();
+	let mut extra_benchmark_names: Vec<Ident> = Vec::new();
+	let mut skip_meta_benchmark_names: Vec<Ident> = Vec::new();
+
+	let (_brace, mut content) =
+		module.content.ok_or(syn::Error::new(mod_span, "Module cannot be empty!"))?;
+
+	// find all function defs marked with #[benchmark]
+	let benchmark_fn_metas = content.iter_mut().filter_map(|stmt| {
+		// parse as a function def first
+		let Item::Fn(func) = stmt else { return None };
+
+		// find #[benchmark] attribute on function def
+		let benchmark_attr = func.attrs.iter().find_map(|attr| {
+			let seg = attr.path.segments.last()?;
+			syn::parse::<keywords::benchmark>(seg.ident.to_token_stream().into()).ok()?;
+			Some(attr)
+		})?;
+
+		Some((benchmark_attr.clone(), func.clone(), stmt))
+	});
+
+	// parse individual benchmark defs and args
+	for (benchmark_attr, func, stmt) in benchmark_fn_metas {
+		// parse any args provided to #[benchmark]
+		let attr_tokens = benchmark_attr.tokens.to_token_stream().into();
+		let benchmark_args: BenchmarkAttrs = syn::parse(attr_tokens)?;
+
+		// parse benchmark def
+		let benchmark_def =
+			BenchmarkDef::from(&func, benchmark_args.extra, benchmark_args.skip_meta)?;
+
+		// record benchmark name
+		let name = &func.sig.ident;
+		benchmark_names.push(name.clone());
+
+		// record name sets
+		if benchmark_def.extra {
+			extra_benchmark_names.push(name.clone());
+		}
+		if benchmark_def.skip_meta {
+			skip_meta_benchmark_names.push(name.clone())
+		}
+
+		// expand benchmark
+		let expanded = expand_benchmark(benchmark_def, name, instance, where_clause.clone());
+
+		// replace original function def with expanded code
+		*stmt = Item::Verbatim(expanded);
+	}
+
+	// generics
+	let type_use_generics = match instance {
+		false => quote!(T),
+		true => quote!(T, I),
+	};
+	let type_impl_generics = match instance {
+		false => quote!(T: Config),
+		true => quote!(T: Config<I>, I: 'static),
+	};
+
+	let krate = generate_crate_access_2018("frame-benchmarking")?;
```

Comment:
You are exporting the macro from `frame-support`, but require here that `frame-benchmarking` is present in the `Cargo.toml`. This is a macro anti pattern. The crate from where you re-export the macro is the crate that you search for in the `Cargo.toml`.

Reply:
Yeah the reason I need `frame-benchmarking` is because this is where all the benchmarking-related traits and structs are. I put things in frame support for two reasons:

1. if I put this in `frame-benchmarking` there would be a name collision with `benchmarks!`
2. `frame-benchmarking` does not have a sub-crate for procedural macros (like the one in `frame_support`), so if we do move this over to `frame-benchmarking`, we will have to deal with the name collision issue.

Open to any suggestions. Discussion would be great to have on the follow-up issue here: https://github.com/paritytech/substrate/issues/13132

Reply:
> 2\. `frame-benchmarking` does not have a sub-crate for procedural macros (like the one in `frame_support`), so if we do move this over to `frame-benchmarking`, we will have to make a sub-crate and deal with the name collision issue.

You can still reuse `frame-support-procedural`, but just re-export the macro from `frame-benchmarking`? 

> 1. if I put this in `frame-benchmarking` there would be a name collision with `benchmarks!`

When you have a declarative macro and a type with the same name, you import both with the same name. Then when you use Rust resolves the correct thing you want. Does that maybe also works for the attribute macro? Or we rename the proc-macro.

Reply:
> Does that maybe also works for the attribute macro? Or we rename the proc-macro.

This is a great suggestion if this is true. I will try it out and see! Then we could have everything live under `frame_benchmarking`

Reply:
update: this seems to work, see #13235